File: solutions/elixir/bob/1/bob.exs
defmodule Teenager do
  def hey(str) do
    cond do
      is_nothing?(str) -> "Fine. Be that way!"
      is_yelling?(str) -> "Woah, chill out!"
      is_asking?(str)  -> "Sure."
      true             -> "Whatever."
    end
  end

  defp is_yelling?(str), do: String.upcase(str) == str
  defp is_asking?(str),  do: String.last(str) == "?"
  defp is_nothing?(str), do: str == nil || String.strip(str) == ""
end

File: solutions/swift/bob/1/Bob.swift
//
//  Bob.swift
//  Bob
//
//  Created by Barrett Clark on 9/5/14.
//  Copyright (c) 2014 Barrett Clark. All rights reserved.
//

import Foundation

class Bob: NSObject {
  class func hey(input: String) -> String {
    let phrase = Conversation(phrase: input)
    if (phrase.isSilent()) {
      return "Fine, be that way."
    } else if (phrase.isShouting()) {
      return "Woah, chill out!"
    } else if (phrase.isAsking()) {
      return "Sure."
    }
    return "Whatever."
  }
}

class Conversation: NSObject {
  var phrase: NSString
  init(phrase: NSString) {
    self.phrase = phrase
    super.init()
  }

  func isShouting() -> Bool {
    let characterSet = NSCharacterSet.uppercaseLetterCharacterSet()
    let rangeOfCharacters = phrase.rangeOfCharacterFromSet(characterSet)
    
    return phrase == phrase.uppercaseString && rangeOfCharacters.location != NSNotFound
  }
  func isAsking() -> Bool {
    return phrase.hasSuffix("?")
  }
  func isSilent() -> Bool {
    let trimmedPhrase = phrase.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceAndNewlineCharacterSet())
    return trimmedPhrase == ""
  }
}